Merge feature-serverside-apply back in master

apelisse opened the pull request:
**What type of PR is this?**
/kind api-change
/kind feature

**Special notes for your reviewer**:

Status of this pull-request:
- CRDs are missing, and that's the main reason for being WIP. We're working on this.
- Merging objects should be working mostly fine (there is a known bug with lists),
- Conflict detection and reporting should work,
- All Update/Create paths update the `managedFields` when the feature is enabled, allowing conflicts detection with all the actors of the system.

Most of the logic lives in `vendor/sigs.k8s.io/structured-merge-diff`, the code strictly in kubernetes is for: 
1. Kubectl changes (kubectl apply and kubectl diff now support server-side apply),
2. Types: we've added the `managedFields` field to manage "ownership".
2. api-machinery to plug all the things together, create the new content-type and use it, update the `managedFields` as needed on all modification paths.

@deads2k @justinsb @liggitt 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
for the moment.

apelisse commented:
cc @jennybuckley @kwiesmueller @lavalamp 

I'm not sure all the verify scripts are passing yet, I'll fix as they fail, maybe (since this is just a wip, I don't want to spend too much time fixing these).

Commit:
Add structured-merge-diff vendor

This is where most of the logic actually lives.

Commit:
API Machinery, Kubectl and tests

Commit:
Random unrelated vendor

apelisse commented:
I've updated the description, I'll fix the build/tests.

justinsb commented on pkg/api/testing/defaulting_test.go:
```diff
@@ -49,7 +49,6 @@ func TestDefaulting(t *testing.T) {
 		{Group: "", Version: "v1", Kind: "ConfigMap"}:                                             {},
 		{Group: "", Version: "v1", Kind: "ConfigMapList"}:                                         {},
 		{Group: "", Version: "v1", Kind: "Endpoints"}:                                             {},
-		{Group: "", Version: "v1", Kind: "EndpointsList"}:                                         {},
```
Deliberate?

lavalamp replied:
Please fix.

fedebongio commented:
/assign @mbohlool

deads2k commented:
How is the effort to bring in CRD handling going?

deads2k commented:
To put @apelisse's mind at ease.  I've had a look and it looks plausible and the feature gate looks good.  My biggest concern is actually the CLI flags to use this since they last for a year.  I'd like to see a separate commit for the CRD integration and after this merges a separate pull that clearly makes the apply flags alpha for sig-cli to review.

smarterclayton commented:
/hold

I actually saw one example (which I hope is accurate) of the actual JSON this generates for the first time today, and I'm a bit taken aback at the complexity and vagueness of the schema.  Here's what I saw:

```
{
   "apiVersion": "v1",
   "kind": "ServiceAccount",
   "metadata": {
       "creationTimestamp": "2019-01-24T16:45:18Z",
       "labels": {
           "addonmanager.kubernetes.io/mode": "Reconcile",
           "kubernetes.io/cluster-service": "true"
       },
       "managedFields": {
           "create-v1@2019-01-24T16:45:18Z": {
               "apiVersion": "v1",
               "fields": {
                   "f:metadata": {
                       "f:labels": {
                           ".": null,
                           "f:addonmanager.kubernetes.io/mode": null,
                           "f:kubernetes.io/cluster-service": null
                       },
                       "f:name": null,
                       "f:namespace": null
                   }
               }
           },
           "update-v1@2019-01-24T16:45:19Z": {
               "apiVersion": "v1",
               "fields": {
                   "f:metadata": {
                       "f:selfLink": null
                   },
                   "f:secrets": {
                       ".": null,
                       "k:{\"name\":\"kube-dns-token-ptntt\"}": {
                           ".": null,
                           "f:name": null
                       }
                   }
               }
           }
       },
       "name": "kube-dns",
       "namespace": "kube-system",
       "resourceVersion": "159",
       "selfLink": "/api/v1/namespaces/kube-system/serviceaccounts/kube-dns",
       "uid": "6eda3ef3-1ff7-11e9-8945-480fcf331b9b"
   },
   "secrets": [
       {
           "name": "kube-dns-token-ptntt"
       }
   ]
}
```

(if this is inaccurate, please point me to something better)

1. The unstructured keys of `managedFields` is a huge problem - we've created a micro schema that everyone will end up following in the name, which is not good
2. The way that special metadata fields (name, self link, namespace) show up in diffs is not only concerning from a clarity perspective, but it makes me question whether anyone will ever understand this.
3. The "create" field is going to be larger than the initial object in many cases - which leads me to ask whether the moment this is enabled whether some objects that are under git ops management won't instantly blow up the etcd object size

Can you point me to some other real world examples of this in use so I can critique?  I looked through the PR and couldn't find a real practical example, so either I missed it or you didn't have one in a test (which we would need)>

liggitt commented:
there's likely a better place to have this discussion than the integration PR, so I'm happy to move the conversation elsewhere, just point me in the right direction.

I agree that some examples would be useful to see:

Appearance after creation:
1. what does a small object (like a serviceaccount) look like after `kubectl create -f ...`
2. what does a small object (like a serviceaccount) look like after `kubectl apply -f ...`
3. what does a big object (like a multi-container deployment with envvars, volumes, etc) look like after `kubectl create -f ...`
4. what does a big object (like a multi-container deployment with envvars, volumes, etc) look like after being created with `kubectl apply -f ...`

Appearance after update by a user:
1. All of the above examples after something like `kubectl label ... foo=bar` is done

Appearance after update by a controller:
1. The above objects both have controllers update them, so seeing the appearance after that is done would be helpful

The two main areas I have questions about are understandability and document size:

For understandability:
* Are we able to filter out metadata fields we know should not be owned/managed? Name and namespace are immutable, selfLink, resourceVersion, creationTimestamp, deletionTimestamp, etc are system-set

For document size, we need to understand the impact in the following situations:
* will an object that was previously over X% of the etcd document limit now exceed that limit and be unable to be created/updated?
* do the timestamped stanza names mean that an object that is regularly updated by a controller can accumulate additional stanzas without bound? are we protecting against that, and how do we handle cases where the managed field info is causing us to exceed max document size?

smarterclayton commented:
Also, is this O(N) with edits?  What is the threshold?  I don't want this to be "complete history of all edits", that is *not* what apply should before for.  Compaction, aging out, etc.

apelisse commented:
> Also, is this O(N) with edits? What is the threshold? I don't want this to be "complete history of all edits", that is not what apply should before for. Compaction, aging out, etc.

It's not O(|edits|), it's O(|fields in the object|). Since when a field is modified, it's going to be stolen from the other managers ownership. 

Initially, we were planning on aggregating all the fields from one manager under 1 dictionary item, but since we can have the ability to create multiple managers, we decided it might be helpful and the O space doesn't change. We can easily remove that if needed.

> I don't want this to be "complete history of all edits"

That's not what it is.

smarterclayton commented:
I remove my hold, we had a good discussion on some immediate tweaks, and will update with next steps.

Commit:
Update smd repo to handle CRDs

Commit:
Plugin FieldManager in CRD handler, change to API

Commit:
Run generation scripts

Commit:
Add object information on FieldManager failures

apelisse commented:
/test pull-kubernetes-bazel-test

apelisse commented:
/test pull-kubernetes-e2e-gce

apelisse commented:
OK Tests are passing, we've updated the API as discussed. We are planning on removing the irrelevant fields in a follow-up PR. We've also fixed the problem with list that was mentioned in the opening comment. PTAL. 

@smarterclayton @deads2k @liggitt @lavalamp

lavalamp commented on pkg/api/pod/util_test.go:
```diff
@@ -176,6 +176,10 @@ func collectResourcePaths(t *testing.T, resourcename string, path *field.Path, n
 	case reflect.Ptr:
 		resourcePaths.Insert(collectResourcePaths(t, resourcename, path, name, tp.Elem()).List()...)
 	case reflect.Struct:
+		// Specifically skip ObjectMeta because it has recursive types
```
This is not a good justification for this skip. If the problem were recursive types, the solution would be cycle detection.

Fortunately, there is a better justification: ObjectMeta is generic and therefore should never have a field with a specific resource's name. (Assuming I understand what this function is doing.)

lavalamp commented on pkg/api/v1/persistentvolume/util_test.go:
```diff
@@ -248,6 +247,9 @@ func collectSecretPaths(t *testing.T, path *field.Path, name string, tp reflect.
 	case reflect.Ptr:
 		secretPaths.Insert(collectSecretPaths(t, path, name, tp.Elem()).List()...)
 	case reflect.Struct:
+		if name == "ObjectMeta" {
```
Add: `// objectMeta should not have any field with the word "secret" in it; it contains cycles so it's easiest to just skip it.`

lavalamp commented on pkg/api/v1/pod/util_test.go:
```diff
@@ -340,6 +340,10 @@ func collectResourcePaths(t *testing.T, resourcename string, path *field.Path, n
 	case reflect.Ptr:
 		resourcePaths.Insert(collectResourcePaths(t, resourcename, path, name, tp.Elem()).List()...)
 	case reflect.Struct:
+		// Specifically skip ObjectMeta because it has recursive types
```
Wow, how many copies of this terrible function do we have?

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/meta.go:
```diff
@@ -168,3 +170,14 @@ func (meta *ObjectMeta) SetOwnerReferences(references []OwnerReference) {
+func (meta *ObjectMeta) SetManagedFields(ManagedFields map[string]VersionedFields) {
+	meta.ManagedFields = make(map[string]VersionedFields, len(ManagedFields))
+	for key, value := range ManagedFields {
+		meta.ManagedFields[key] = value
```
Is it OK that this doesn't do a deep copy? If a shallow copy is OK why is it necessary to copy at all?

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1009,3 +1042,22 @@ const (
+// Fields stores a set of fields in a data structure like a Trie.
+// To understand how this is used, see: https://github.com/kubernetes-sigs/structured-merge-diff
+type Fields struct {
+	// Map stores a set of fields in a data structure like a Trie.
```
What is the key format?

jennybuckley replied:
// Map stores a set of fields in a data structure like a Trie.
//
// Each key is either a '.' representing the field itself, and will always map to an empty set,
// or a string representing a sub-field or item. The string will follow one of these four formats:
// 'f:<name>', where <name> is the name of a field in a struct, or key in a map
// 'v:<value>', where <value> is the exact json formatted value of a list item
// 'i:<index>', where <index> is position of a item in a list
// 'k:<keys>',  where <keys> is a map of  a list item's key fields to their unique values
//
// If a key maps to an empty Fields value, the field that key represents is part of the set.

Although, the format is probably best defined by https://github.com/kubernetes/kubernetes/blob/feature-serverside-apply/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/pathelement.go

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -133,6 +133,20 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
+
+			obj, err = scope.FieldManager.Update(liveObj, obj, "create")
+			if err != nil {
+				scope.err(fmt.Errorf("failed to update object managed fields: %v", err), w, req)
```
Until we've exhaustively tested using every field from every schema, I don't think we can make this error fail a request. Alpha features mustn't break existing behavior.

lavalamp reviewed (commented):
(I'm not even halfway through the first commit. Want to get these published before I get distracted. Will review the rest asap.)

Commit:
Avoid closure

lavalamp reviewed (commented):
2/3rds of the way through the first commit. locking in my progress so far.

jennybuckley commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -0,0 +1,174 @@
+// object.
+func (f *FieldManager) Update(liveObj, newObj runtime.Object, manager string) (runtime.Object, error) {
+	managed, err := internal.DecodeObjectManagedFields(newObj)
+	// If the managed field is empty or we failed to decode it,
```
This is to protect against clients who don't understand the managedFields field, and strip it from the object

jennybuckley replied:
Also, if someone needs to update the managedFields directly

jennybuckley commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields_test.go:
```diff
@@ -0,0 +1,143 @@
+)
+
+// TestRoundTripManagedFields will roundtrip ManagedFields from the format used by
+// sigs.k8s.io/structured-merge-diff to the wire format (api format) and back
```
This actually is the wire format, the comment is wrong

lavalamp replied:
(please fix comment then :) )

jennybuckley commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -0,0 +1,174 @@
+		return nil, fmt.Errorf("failed to decode managed fields: %v", err)
+	}
+	// We can assume that patchObj is already on the proper version:
+	// it shouldn't have to be converted so that it's not defaulted.
```
Later we do `f.typeConverter.YAMLToTyped(patch)` which gets the version from the patchbytes, and will fail when we attempt to merge the patch into the live object here:
https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/typed/typed.go#L171

We could definitely put in a more explicit check though

Commit:
Fix SubjectAccessReview Validation with ManagedFields

jennybuckley commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go:
```diff
@@ -795,6 +795,9 @@ func TestPatchWithVersionConflictThenAdmissionFailure(t *testing.T) {
 	tc.Run(t)
 }
 
+// TODO: Add test case for "apply with existing uid" verify it gives a conflict error,
```
Yes, we have a test for this in the integration test, https://github.com/kubernetes/kubernetes/pull/72947/files/35c00c46ab56d9742e59eb1fbcd35c3eeed76f81..8178b927ded8fbff0a9d5d70da2f86f8a4aa3daa#diff-4ac5831d494b1b52c7c7be81e552a458R92

I'll remove this TODO

jennybuckley commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -0,0 +1,174 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert live object to proper version: %v", err)
+	}
+	if err := internal.RemoveObjectManagedFields(liveObjVersioned); err != nil {
```
We hit an error before when assuming an object will have an ObjectMeta accessor, (which is the way RemoveObjectManagedFields can fail) since DeploymentRollback didn't have an ObjectMeta

lavalamp reviewed (commented):
All the way through the first commit! \o/

lavalamp commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:
```diff
@@ -21,6 +21,18 @@ import (
 	"net/http"
 	"time"
 
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
```
I won't ask for a change but making large reorganizations to import lists doesn't make this PR any smaller :)

apelisse replied:
But ... there was an extra empty line that shouldn't be there? :-)

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -60,6 +62,22 @@ func NewFieldManager(models openapiproto.Models, objectConverter runtime.ObjectC
 
+// NewCRDFieldManager creates a new FieldManager specifically for
+// CRDs. This doesn't use openapi models (and it doesn't support the
+// validation field right now).
```
Are we keeping a list somewhere of things that need to be fixed?

apelisse replied:
This list is mostly the spreadsheet that we used to have. As this gets released to alpha and we work on promoting to beta, I think it'd make sense to move some of the items that we have (and create the new items) in a github umbrella issue for transparency, etc.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/managedfields.go:
```diff
@@ -72,21 +74,53 @@ func EncodeObjectManagedFields(obj runtime.Object, fields fieldpath.ManagedField
 
-func decodeVersionedSet(encodedVersionedSet *metav1.VersionedFields) (versionedSet *fieldpath.VersionedSet, err error) {
+// DecodeManager creates a manager identifier string from a ManagedFieldsEntry
+func DecodeManager(encodedManager *metav1.ManagedFieldsEntry) (manager string, err error) {
```
Please make this function name more accurate?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/typeconverter.go:
```diff
@@ -21,21 +21,52 @@ import (
+// DeducedTypeConverter is a TypeConverter for CRDs that don't have a
+// schema. It does implement the same interface though (and create the
+// same types of objects), so that everything can still work the same.
+// CRDs are merged with all their fields being "atomic" (lists
```
This is maybe OK for alpha but I think we need something better for beta and up, can we add a todo or something else to this comment?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go:
```diff
@@ -191,3 +192,7 @@ type namedCreaterAdapter struct {
 }
+
+func prefixFromUserAgent(u string) string {
+	return strings.Split(u, "/")[0]
```
We're sure this is useful and doesn't return "kubernetes" for everything?

apelisse replied:
Some examples of values we've seen:
`kubelet`, `hyperkube`, `kubectl`, ...

lavalamp reviewed (commented):
OK I think this covers everything else.

jennybuckley commented on test/integration/auth/rbac_test.go:
```diff
@@ -478,6 +482,40 @@ func TestRBAC(t *testing.T) {
 				{"limitrange-updater", "PUT", "", "limitranges", "limitrange-namespace", "a", aLimitRange, http.StatusOK},
 			},
 		},
+		// {
```
I just uncommitted it and everything passes

Commit:
Minor fixes

lavalamp commented:
I reviewed the smd changes, too-- I think we should make some changes when we start using the schemas from CRDs but that doesn't need to block this.

What else do we want to do before merging?

apelisse commented:
This is it.

lavalamp commented:
OK, great. It's essential that we update the KEP ASAP, I think. I thought I would do it but I haven't had time to even open it up yet.

lavalamp commented:
Let's merge on Monday morning so we'll all be around in the super unlikely event something goes wrong :)

kwiesmueller commented:
Here is the WIP follow-up PR about stripping certain fields from the managedFields.
https://github.com/kubernetes/kubernetes/pull/73681

cc. @liggitt

deads2k commented on vendor/sigs.k8s.io/structured-merge-diff/fieldpath/path.go:
```diff
@@ -35,6 +35,12 @@ func (fp Path) String() string {
 	return strings.Join(strs, "")
 }
 
+func (fp Path) Copy() Path {
```
nit: Because the diff is small I can chase refs for looking at shallow copy handling of each element, but future generations may appreciate a comment

deads2k commented:
> Let's merge on Monday morning so we'll all be around in the super unlikely event something goes wrong :)

I'm good with this.  The CRD integration was surprisingly clean.  Nice job.

apelisse commented:
> Thanks @jennybuckley & @apelisse for this! This is super exciting.

:sunglasses:

lavalamp commented:
@smarterclayton I just saw your comment which the gtihub UI had helpfully hidden until now. Have the reasons for your hold been addressed?

lavalamp commented:
/hold cancel

Ah, @smarterclayton seems happy: https://github.com/kubernetes/kubernetes/pull/72947#issuecomment-457317378

apelisse commented:
/priority important-soon

lavalamp commented:
/lgtm

tedyu commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go:
```diff
@@ -0,0 +1,225 @@
+	// let's try the live object. This is to prevent clients who
+	// don't understand managedFields from deleting it accidentally.
+	if err != nil || len(managed) == 0 {
+		managed, err = internal.DecodeObjectManagedFields(liveObj)
```
Is it possible that managed is empty and err == nil after the call is made ?

tedyu commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/gvkparser.go:
```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
It would be nice if year is updated.

tedyu commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/gvkparser.go:
```diff
@@ -0,0 +1,116 @@
+	for _, modelName := range models.ListModels() {
+		model := models.LookupModel(modelName)
+		if model == nil {
+			panic("ListModels returns a model that can't be looked-up.")
```
Please include modelName in the message

tedyu commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/internal/gvkparser.go:
```diff
@@ -0,0 +1,116 @@
+		gvkList := parseGroupVersionKind(model)
+		for _, gvk := range gvkList {
+			if len(gvk.Kind) > 0 {
+				parser.gvks[gvk] = modelName
```
What if parser.gvks[gvk] already had some value ?

jennybuckley commented:
@tedyu Thanks for the review, we will address these comments in a follow-up PR, since this one is about to merge.

tedyu commented:
@jennybuckley 
I can send PR for the points I made, if you're busy.

Cheers

tedyu commented:
@jennybuckley 
Please see #73731.
For my last comment, once we decide what to do with duplicate gvk key, I can add it to the PR.

felixfbecker commented on staging/src/k8s.io/apimachinery/pkg/types/patch.go:
```diff
@@ -25,4 +25,5 @@ const (
 	JSONPatchType           PatchType = "application/json-patch+json"
 	MergePatchType          PatchType = "application/merge-patch+json"
 	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
+	ApplyPatchType          PatchType = "application/apply-patch+yaml"
```
Will `application/apply-patch+json` be supported too?

apelisse replied:
Good question, we've decided not to do that because json is valid yaml. Feel free to send json directly to this end-point.

apelisse replied:
Adding a comment there might be helpful.

jennybuckley commented:
/wg apply
